bootstrap/controllers/machine_deployment_controller.go

```diff
@@ -0,0 +1,433 @@
+	}
+
+	// clean up
+	delete(annotations, bootstrapv1.InPlaceUpgradeReleaseAnnotation)
+
+	annotations[bootstrapv1.InPlaceUpgradeToAnnotation] = scope.UpgradeTo
+	annotations[bootstrapv1.InPlaceUpgradeStatusAnnotation] = bootstrapv1.InPlaceUpgradeInProgressStatus
+
+	if err := scope.PatchHelper.Patch(ctx, scope.MachineDeployment); err != nil {
+		return fmt.Errorf("failed to patch: %w", err)
+	}
+
+	r.recorder.Eventf(
+		scope.MachineDeployment,
+		corev1.EventTypeNormal,
+		bootstrapv1.InPlaceUpgradeInProgressEvent,
+		"In-place upgrade is in-progress",
+		scope.MachineDeployment.Name,
+	)
+	return nil
+}
+
+// markUpgradeDone marks the MachineDeployment as in-place upgrade done.
+func (r *MachineDeploymentReconciler) markUpgradeDone(ctx context.Context, scope *MachineDeploymentUpgradeScope) error {
+	annotations := scope.MachineDeployment.Annotations
+	if annotations == nil {
+		annotations = make(map[string]string)
+	}
+
+	// clean up
+	delete(annotations, bootstrapv1.InPlaceUpgradeToAnnotation)
+
+	annotations[bootstrapv1.InPlaceUpgradeStatusAnnotation] = bootstrapv1.InPlaceUpgradeDoneStatus
+	annotations[bootstrapv1.InPlaceUpgradeReleaseAnnotation] = scope.UpgradeTo
+
+	if err := scope.PatchHelper.Patch(ctx, scope.MachineDeployment); err != nil {
+		return fmt.Errorf("failed to patch: %w", err)
+	}
+
+	r.recorder.Eventf(
+		scope.MachineDeployment,
+		corev1.EventTypeNormal,
+		bootstrapv1.InPlaceUpgradeDoneEvent,
+		"In-place upgrade is done",
+		scope.MachineDeployment.Name,
+	)
+	return nil
+}
+
+// markUpgradeFailed marks the MachineDeployment as in-place upgrade failed.
+func (r *MachineDeploymentReconciler) markUpgradeFailed(ctx context.Context, scope *MachineDeploymentUpgradeScope, failedM *clusterv1.Machine) error {
+	annotations := scope.MachineDeployment.Annotations
+	if annotations == nil {
+		annotations = make(map[string]string)
+	}
+
+	// clean up
+	delete(annotations, bootstrapv1.InPlaceUpgradeReleaseAnnotation)
+
+	annotations[bootstrapv1.InPlaceUpgradeStatusAnnotation] = bootstrapv1.InPlaceUpgradeFailedStatus
+
+	if err := scope.PatchHelper.Patch(ctx, scope.MachineDeployment); err != nil {
+		return fmt.Errorf("failed to patch: %w", err)
+	}
+
+	r.recorder.Eventf(
+		scope.MachineDeployment,
+		corev1.EventTypeWarning,
+		bootstrapv1.InPlaceUpgradeFailedEvent,
+		"In-place upgrade failed for machine %q.",
+		failedM.Name,
+	)
+	return nil
+}
+
+// createScope creates a new MachineDeploymentUpgradeScope.
+func (r *MachineDeploymentReconciler) createScope(md *clusterv1.MachineDeployment, ownedMachines []*clusterv1.Machine) (*MachineDeploymentUpgradeScope, error) {
+	patchHelper, err := patch.NewHelper(md, r.Client)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create new patch helper: %w", err)
+	}
+
+	// NOTE(Hue): The reason we are checking the `Release` annotation as well is that we want to make sure
+	// we upgrade the new machines that joined after the initial upgrade process.
```

Comment:
I'm not sure what's the intention here, the `CK8sConfigReconciler` checks for the release annotation and overrides the install already. I believe just propagating the release annotation should be enough.

Reply:
Oh I was not aware of this: https://github.com/canonical/cluster-api-k8s/blob/main/bootstrap/controllers/ck8sconfig_controller.go#L400
Thanks for pointing this out.